TrackletAlgorithm/MemoryTemplateBinned.h

```diff
@@ -21,7 +21,7 @@ class MemoryTemplateBinned{
 public:
   static constexpr unsigned int kNBitDataAddr = NBIT_ADDR-NBIT_BIN;
   typedef ap_uint<NBIT_BX> BunchXingT;
-  typedef ap_uint<kNBitDataAddr> NEntryT;
+  typedef ap_uint<kNBitDataAddr+1> NEntryT;
```

Comment:
does this change need to be discussed? It's the only one outside the VMR files, so it might affect other people?

Reply:
Discussed it with @djcranshaw, but don't know if more people need to be involved? I think it had the +1 previously, but it was removed one day which then caused errors when test vectors contained more stubs in a bin than the max limit.

Reply:
So the reason that NEntryT for binned memories needs 5 bits instead of 4 is that a particular bin can have anywhere from 0-16 entries, which means 17 possibilities. This isn't true of unbinned memories which can only have up to 108 out of 128 entries, and so only 7 bits for NEntryT are needed. It's true that any module which reads a binned memory (which means TE and ME) could be affected in principle by this change, although if they are using the MemoryTemplateBinned::NEntryT type (or auto) inside their module when reading the number of entries, then no actual code changes should be needed.

Reply:
does that mean `kNBitDataAdd` should get the +1?

Reply:
I don't think so. Pragmatically it is convenient if the kNBitDataAddr remains 4 because when looping over the entries in a bin (in e.g. print_mem() or clear()) the iteration variable needs only to range from 0-15, and in write_mem(), we want the write to succeed only if the attempted write address is 0-15. And philosophically, it remains true that we use only 4 bits to represent the address of the slots in a bin. It is only because the number of entries has to be able to represent a completely empty and a completely full bin that it needs one extra bit, so I think the +1 should stay with NEntryT.

Reply:
ok.  I think I understand now. Looking at how the variables are used I see a lot of `1<<(kNBitDataAddr)` in the code; maybe we can use this opportunity to clean that up.

Reply:
>  It's true that any module which reads a binned memory (which means TE and ME) could be affected in principle by this change, although if they are using the MemoryTemplateBinned::NEntryT type (or auto) inside their module when reading the number of entries, then no actual code changes should be needed.

@djcranshaw can you do this check. I was going to try to check but I realize I'm not sure I understand.

Reply:
Sure, I'll check both modules.

Reply:
Both ME and TE use auto const when retrieving nEntries for binned memories (L131,L132 of TrackletEngine.h and L110,L111 of MatchEngine.cc), so no changes should be needed in these modules.